src/basic/macro.h

```diff
@@ -449,4 +449,61 @@ assert_cc(sizeof(dummy_t) == 0);
                 _b_;                                                    \
         })
 
+/* Bit index(es) (0-based) to mask (of specified type). Assertion failure if (any) index is out of range. */
+#define INDEXES_TO_MASK(type, ...) _INDEXES_TO_MASK(type, UNIQ_T(_mask_, UNIQ), UNIQ_T(_indexes_, UNIQ), UNIQ_T(_i_, UNIQ), ##__VA_ARGS__)
+#define __INDEXES_TO_MASK(type, _mask_, _indexes_, _i_, ...)             \
+        ({                                                              \
+                type _mask_ = 0;                                        \
+                size_t _indexes_[] = { __VA_ARGS__ };                   \
+                for (size_t _i_ = 0; _i_ < ELEMENTSOF(_indexes_); _i_++) \
+                        _mask_ |= INDEX_TO_MASK(type, _indexes_[_i_];   \
+                _mask_;                                                 \
+        })
+#define _INDEXES_TO_MASK(type, _mask_, _indexes_, _i_, ...)             \
+        ({                                                              \
+                type _mask_ = 0;                                        \
+                VA_ARGS_FOREACH(_i_, size_t, __VA_ARGS__)               \
+                        _mask_ |= INDEX_TO_MASK(type, _i_);             \
+                _mask_;                                                 \
+        })
+#define INDEX_TO_MASK(type, index) _INDEX_TO_MASK(type, index, UNIQ_T(_index_, UNIQ))
+#define _INDEX_TO_MASK(type, index, _index_)                            \
+        ({                                                              \
+                size_t _index_ = (index);                               \
+                assert(sizeof(type) <= sizeof(unsigned long long));     \
+                assert(_index_ < sizeof(type) * 8);                     \
+                ((type)1) << _index_;                                   \
+        })
+
+/* Set, clear, or check bit. Index is 0-based. Assertion failure if index is out of range. */
+#define SET_BIT(bits, index) ((bits) |= INDEX_TO_MASK(typeof(bits), index))
+#define CLEAR_BIT(bits, index) ((bits) &= ~INDEX_TO_MASK(typeof(bits), index))
+#define IS_BIT_SET(bits, index) (!!((bits) & INDEX_TO_MASK(typeof(bits), index)))
```

Comment:
so these we kinda have already though differently, via SET_FLAG() + FLAGS_SET(). they don't do the same thing, but having them all seems a bit redundant. At the very least, they should reuse each other.

Reply:
ack, will update to use the flags macros